### PR TITLE
ci: Run CI against Symfony 5 and 6

### PR DESCRIPTION
**Pull request by sidz**
This PR improves our pipeline and running CI with Symfony 5 and 6.

### CODE REVIEW

File: .github/workflows/autoreview.yaml [hunk 1]
```diff
@@ -22,20 +22,20 @@ jobs:
 
     steps:
       - name: Checkout code
-        uses: actions/checkout@v2
```
Comment (theofidry) [inline .github/workflows/autoreview.yaml:25]: at the very least those changes look safe enough to be extracted to reduce the noise?

Comment (sidz) [reply .github/workflows/autoreview.yaml:25]: yeah. Going to revert them

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -55,23 +65,32 @@ jobs:
...
             composer-
 
-      - name: Install dependencies
+      - name: Install locked dependencies
+        if: ${{ matrix.dependencies == 'locked' }}
         run: |
-          composer install --no-interaction --prefer-dist --no-progress
+          composer install --no-interaction --prefer-dist --no-progress --optimize-autoloader
+
+      - name: Install lowest dependencies
```
Comment (theofidry) [inline .github/workflows/ci.yaml:80]: I think we could migrate to the ramsey action which offers the flexibility to set lowest/highest, to do the update or install and also offers caching. This is nothing that we can't do, but that would remove some boiler-template and maybe make this PR a bit easier

File: .github/workflows/mt.yaml [hunk 3]
```diff
@@ -53,7 +53,7 @@ jobs:
 
       - name: Install dependencies
         run: |
-          composer update --no-interaction --prefer-dist --no-progress
+          composer install --no-interaction --prefer-dist --no-progress
```
Comment (sidz) [inline .github/workflows/mt.yaml:56]: I don't remember why do we use "highest" dependencies instead of locked. So lets use locked

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -17,28 +17,33 @@ jobs:
...
-        php-version: [ '8.0' ]
+        php-version: [ '8.0', '8.1', '8.2' ]
         coverage-driver: [ pcov, xdebug ]
-        e2e-runner: [ 'bin/infection' ]
+        e2e-runner: [ 'bin/infection', 'build/infection.phar' ]
+        dependencies:
+          - "locked"
+          - "lowest"
+          - "highest"
+        exclude:
```
Comment (sidz) [inline .github/workflows/ci.yaml:27]: we probably may exclude more cases. Give me nudge if we need to add something else

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -55,23 +60,32 @@ jobs:
...
+        run: |
+          composer update --no-interaction --prefer-dist --no-progress --optimize-autoloader
 
       - name: Run tests
         shell: bash
         run: |
-          if [[ "${{ matrix.php-version }}" == '8.2' ]]; then
-            export SYMFONY_DEPRECATIONS_HELPER=max[self]=0
-          fi
+          export SYMFONY_DEPRECATIONS_HELPER=max[self]=0
```
Comment (sidz) [inline .github/workflows/ci.yaml:88]: We will need to remove this hack and fix all direct deprecations. For instance `strlen`

File: .github/workflows/mt.yaml (hunk 3 above)
Comment (theofidry) [reply .github/workflows/mt.yaml:56]: probably because then we never test infection installed as a package, since it will only picked the locked with the configured platform

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (theofidry) [inline .github/workflows/ci.yaml:88]: Maybe we can sort it out in a separate PR, but I wonder if we really want to enable the deprecations when not running the locked dependencies

File: psalm.xml [hunk 6]
```diff
@@ -4,7 +4,7 @@
     resolveFromConfigFile="true"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xmlns="https://getpsalm.org/schema/config"
-    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
+    xsi:schemaLocation="https://getpsalm.org/schema/config.xsd vendor/vimeo/psalm/config.xsd"
```
Comment (theofidry) [inline psalm.xml:7]:
```suggestion
    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
```

**Comment by sidz** [comment]
Currently I'm waiting when @maks-rafalko adds GITHUB_TOKEN secret in the repository settings to avoid rate limit exception

**Comment by theofidry** [comment]
@sidz actually the `GITHUB_TOKEN` secret should already be configured: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#about-the-github_token-secret

**Comment by sidz** [comment]
thanks @theofidry for pointing to the documentation. I assumed that secret needs to be added manually

**Comment by theofidry** [comment]
@sidz I went ahead with my refactors. Just to me clear I'm avoiding adding the Symfony6 support to not overlap with your work. If this hinders this PR by resulting in conflicts I'll do a PR on yours to resolve them.

File: .github/workflows/mt.yaml (hunk 3 above)
Comment (theofidry) [reply .github/workflows/mt.yaml:59]: Sorted out in #1791

File: composer.json [hunk 7]
```diff
@@ -88,9 +88,6 @@
         "thecodingmachine/phpstan-safe-rule": "^1.2.0"
     },
     "config": {
-        "platform": {
```
Comment (theofidry) [inline composer.json:91]: I don't think removing the platform completely is the way, since we do want it. More likely we want to unset it when testing non-locked dependencies like done in https://github.com/infection/infection/pull/1786/files#diff-6e608e02c595d53ab6b70822a2bf19abcfc6ddcc976c2f536ad5bfca20f0443fR51-R53

File: tests/phpunit/Fixtures/Finder/MockRealPathFinder.php [hunk 8]
```diff
@@ -1,21 +1,20 @@
 <?php
 
+declare(strict_types=1);
```
Comment (theofidry) [inline tests/phpunit/Fixtures/Finder/MockRealPathFinder.php:3]: Looks like we could extract those changes in a different PR?

File: tests/phpunit/Fixtures/Finder/MockRelativePathFinder.php [hunk 9]
```diff
@@ -1,21 +1,20 @@
 <?php
 
+declare(strict_types=1);
```
Comment (theofidry) [inline tests/phpunit/Fixtures/Finder/MockRelativePathFinder.php:3]: same for this one

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (theofidry) [reply .github/workflows/ci.yaml:27]: For me it's quite hard to tell at this point, because I don't know what is changed here because we were not testing a wide enough set, or because of the Symfony6 compatibility.

Don't you think it would be best to go towards #1786 first, which allows to test a wider range whilst keeping the Symfony6 compatibility issue out, and then add Symfony6 to fix the compatibility?

If the conflicts are a concern I'm more than happy to take care of them

**Review by theofidry — COMMENTED** [review]
👍 overall. I think there is a bit too much mixed in there for my taste to distinguish what is due to Symfony6 and what is due to test a wider range of combinations.

But if you think it's counter-productive to go with my suggestion I can do my rework later.

**Comment by theofidry** [comment]
@maks-rafalko to me the failure looks legit, but I don't think it's anything new. It just looks like we never ran the unit tests on Windows.

Is it something we care about?
yes -> I would disable it in this PR, create another PR where I enable it and try to fix it separately; since this is not new
no -> we remove it and go on

**Comment by theofidry** [comment]
Notes for myself for when I come back to this:

- Make easier to adapt the protection rules (#1806)
- Split the CI to make it easier (#1787)
- Prepare the CI to test against different dependency versions (#1815)
- Prepare the CI to test against different Symfony versions (#1817)
- Update this PR to add Symfony6 support
- Fixing the incomplete builds:
  - add the missing bits in the range (test against lowest & highest for example)
  - fix the broken builds, e.g. #1805
